### PR TITLE
Interaction with ground

### PR DESCRIPTION
**Pull request by Mol2017**
1. drone stops moving when height is less than 0.05
2. drone stops rotating and has a normal (0, 0, 1)

### CODE REVIEW

File: gym_art/quadrotor_multi/quadrotor_multi.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def __init__(self,
                  viz_traces=25, viz_trace_nth_step=1):
 
         super().__init__()
-
+
```
Comment (Zhehui-Huang) [inline gym_art/quadrotor_multi/quadrotor_multi.py]: Please remove this change.

File: gym_art/quadrotor_multi/quadrotor_single.py [hunk 2]
```diff
@@ -312,7 +312,7 @@ def step1(self, thrust_cmds, dt, thrust_noise):
         thrusts = self.thrust_max * self.angvel2thrust(self.thrust_cmds_damp, linearity=self.motor_linearity)
         # Prop crossproduct give torque directions
         self.torques = self.prop_crossproducts * thrusts[:, None]  # (4,3)=(props, xyz)
-
+
```
Comment (Zhehui-Huang) [inline gym_art/quadrotor_multi/quadrotor_single.py]: Please remove this change.

File: gym_art/quadrotor_multi/quadrotor_single.py [hunk 3]
```diff
@@ -370,6 +370,11 @@ def step1(self, thrust_cmds, dt, thrust_noise):
         ## (Square) Damping using torques (in case we would like to add damping using torques)
         # damping_torque = - 0.3 * self.omega * np.fabs(self.omega)
         self.torque = thrust_torque + rotor_visc_torque
+
+        # Add rotation torque if the drones is on the ground
+        # if self.pos[2] == 0:
+        #     self.torque += - self.omega * self.inertia * 1 / dt
+
```
Comment (Zhehui-Huang) [inline gym_art/quadrotor_multi/quadrotor_single.py]: Please remove this change.

File: gym_art/quadrotor_multi/quadrotor_single.py [hunk 4]
```diff
@@ -413,6 +418,10 @@ def step1(self, thrust_cmds, dt, thrust_noise):
         omega_damp_quadratic = np.clip(self.damp_omega_quadratic * self.omega ** 2, a_min=0.0, a_max=1.0)
         self.omega = self.omega + (1.0 - omega_damp_quadratic) * dt * self.omega_dot
         self.omega = np.clip(self.omega, a_min=-self.omega_max, a_max=self.omega_max)
+        
+        if self.pos[2] <= 0.:
```
Comment (Zhehui-Huang) [inline gym_art/quadrotor_multi/quadrotor_single.py]: self.pos[2] should not be <=0.0, it should be <=0.05.

File: gym_art/quadrotor_multi/quadrotor_single.py [hunk 5]
```diff
@@ -435,6 +444,40 @@ def step1(self, thrust_cmds, dt, thrust_noise):
...
+            else:
+                # step 2.2: rolling friction
+                # check if stops with dt
+                rolling_friction = - vel_xy_dir * F_n * rolling_mu_k
+                acc += rolling_friction / self.mass
+                acc_xy = np.array([acc[0], acc[1], 0.])
+                pred_vel_xy = vel_xy + dt * acc_xy
+                if np.dot(vel_xy, pred_vel_xy) <= 0:
+                    acc[0:2] = 0.
+                    self.vel[0:2] = 0.
```
Comment (Zhehui-Huang) [inline gym_art/quadrotor_multi/quadrotor_single.py]: If pos_z <0.05, should force pos_z = 0.05

**Review by Zhehui-Huang — CHANGES_REQUESTED** [review]

File: gym_art/quadrotor_multi/quadrotor_multi.py (hunk 1 above)
Comment (Mol2017) [reply gym_art/quadrotor_multi/quadrotor_multi.py]: FIxed.

File: gym_art/quadrotor_multi/quadrotor_single.py (hunk 2 above)
Comment (Mol2017) [reply gym_art/quadrotor_multi/quadrotor_single.py]: Fixed

File: gym_art/quadrotor_multi/quadrotor_single.py (hunk 3 above)
Comment (Mol2017) [reply gym_art/quadrotor_multi/quadrotor_single.py]: Unused comments removed

File: gym_art/quadrotor_multi/quadrotor_single.py (hunk 4 above)
Comment (Mol2017) [reply gym_art/quadrotor_multi/quadrotor_single.py]: Fixed

File: gym_art/quadrotor_multi/quadrotor_single.py (hunk 5 above)
Comment (Mol2017) [reply gym_art/quadrotor_multi/quadrotor_single.py]: Fixed, position z is now clamped to [0.05, room_box limit].

**Pull request closed without merging**